src/wasm32_simd.rs

```diff
@@ -0,0 +1,787 @@
+    t2 = shuffle_epi32::<1, 3, 2, 0>(tt);
+    g1(row0, row1, row2, row3, t2);
+    t3 = unpackhi_epi32(m1, m3);
+    tt = unpacklo_epi32(m2, t3);
+    t3 = shuffle_epi32::<0, 1, 3, 2>(tt);
+    g2(row0, row1, row2, row3, t3);
+    undiagonalize(row0, row2, row3);
+    m0 = t0;
+    m1 = t1;
+    m2 = t2;
+    m3 = t3;
+
+    // Round 5
+    t0 = shuffle!(m0, m1, 3, 1, 1, 2);
+    t0 = shuffle_epi32::<0, 3, 2, 1>(t0);
+    g1(row0, row1, row2, row3, t0);
+    t1 = shuffle!(m2, m3, 3, 3, 2, 2);
+    tt = shuffle_epi32::<0, 0, 3, 3>(m0);
+    t1 = blend_epi16(tt, t1, 0xCC);
+    g2(row0, row1, row2, row3, t1);
+    diagonalize(row0, row2, row3);
+    t2 = unpacklo_epi64(m3, m1);
+    tt = blend_epi16(t2, m2, 0xC0);
+    t2 = shuffle_epi32::<1, 3, 2, 0>(tt);
+    g1(row0, row1, row2, row3, t2);
+    t3 = unpackhi_epi32(m1, m3);
+    tt = unpacklo_epi32(m2, t3);
+    t3 = shuffle_epi32::<0, 1, 3, 2>(tt);
+    g2(row0, row1, row2, row3, t3);
+    undiagonalize(row0, row2, row3);
+    m0 = t0;
+    m1 = t1;
+    m2 = t2;
+    m3 = t3;
+
+    // Round 6
+    t0 = shuffle!(m0, m1, 3, 1, 1, 2);
+    t0 = shuffle_epi32::<0, 3, 2, 1>(t0);
+    g1(row0, row1, row2, row3, t0);
+    t1 = shuffle!(m2, m3, 3, 3, 2, 2);
+    tt = shuffle_epi32::<0, 0, 3, 3>(m0);
+    t1 = blend_epi16(tt, t1, 0xCC);
+    g2(row0, row1, row2, row3, t1);
+    diagonalize(row0, row2, row3);
+    t2 = unpacklo_epi64(m3, m1);
+    tt = blend_epi16(t2, m2, 0xC0);
+    t2 = shuffle_epi32::<1, 3, 2, 0>(tt);
+    g1(row0, row1, row2, row3, t2);
+    t3 = unpackhi_epi32(m1, m3);
+    tt = unpacklo_epi32(m2, t3);
+    t3 = shuffle_epi32::<0, 1, 3, 2>(tt);
+    g2(row0, row1, row2, row3, t3);
+    undiagonalize(row0, row2, row3);
+    m0 = t0;
+    m1 = t1;
+    m2 = t2;
+    m3 = t3;
+
+    // Round 7
+    t0 = shuffle!(m0, m1, 3, 1, 1, 2);
+    t0 = shuffle_epi32::<0, 3, 2, 1>(t0);
+    g1(row0, row1, row2, row3, t0);
+    t1 = shuffle!(m2, m3, 3, 3, 2, 2);
+    tt = shuffle_epi32::<0, 0, 3, 3>(m0);
+    t1 = blend_epi16(tt, t1, 0xCC);
+    g2(row0, row1, row2, row3, t1);
+    diagonalize(row0, row2, row3);
+    t2 = unpacklo_epi64(m3, m1);
+    tt = blend_epi16(t2, m2, 0xC0);
+    t2 = shuffle_epi32::<1, 3, 2, 0>(tt);
+    g1(row0, row1, row2, row3, t2);
+    t3 = unpackhi_epi32(m1, m3);
+    tt = unpacklo_epi32(m2, t3);
+    t3 = shuffle_epi32::<0, 1, 3, 2>(tt);
+    g2(row0, row1, row2, row3, t3);
+    undiagonalize(row0, row2, row3);
+
+    [*row0, *row1, *row2, *row3]
+}
+
+#[target_feature(enable = "simd128")]
+pub unsafe fn compress_in_place(
+    cv: &mut CVWords,
+    block: &[u8; BLOCK_LEN],
+    block_len: u8,
+    counter: u64,
+    flags: u8,
+) {
+    let [row0, row1, row2, row3] = compress_pre(cv, block, block_len, counter, flags);
+    // it stores in reversed order...
+    storeu(xor(row0, row2), cv.as_mut_ptr().add(0) as *mut u8);
+    storeu(xor(row1, row3), cv.as_mut_ptr().add(4) as *mut u8);
+}
+
+#[target_feature(enable = "simd128")]
+pub unsafe fn compress_xof(
```

Comment:
Similar to above, the x86 version of these fundamentally can never be safe, because you must not execute new instructions on old chips. But it seems like the Wasm versions can be safe here, as long as the arguments are regular safe reference types and stuff, because they're guaranteed to crash cleanly if support isn't there?

Reply:
Unfortunately, stores and loads are still unsafe, so this function is not safe.